pkg/kubectl/cmd/top.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"errors"
+	"io"
+	"strings"
+
+	"github.com/renstrom/dedent"
+	"github.com/spf13/cobra"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/metricsutil"
+)
+
+// TopOptions contains all the options for running the top cli command.
+type TopOptions struct {
+	ResourceType    unversioned.GroupKind
+	ResourceName    string
+	AllNamespaces   bool
+	PrintContainers bool
+	Selector        string
+	Namespace       string
+	Client          *metricsutil.HeapsterMetricsClient
+	Printer         *metricsutil.TopCmdPrinter
+}
+
+var (
+	topLong = dedent.Dedent(`
+		Display Resource (CPU/Memory/Storage) usage of nodes or pods.
+
+		The top command allows you to see the resource consumption of the nodes or pods.
+		It downloads the usage metrics of a given resource (node/pod) via the Resource Metrics API.`)
+
+	topExample = dedent.Dedent(`
+		  # Show metrics for all nodes
+		  kubectl top nodes
+
+		  # Show metrics for a given node
+		  kubectl top node NODE_NAME
+
+		  # Show metrics for all pods in the given namespace
+		  kubectl top pods --namespace=NAMESPACE
+
+		  # Show metrics for a given pod and its containers
+		  kubectl top pod POD_NAME --containers
+
+		  # Show metrics for the pods defined by label name=myLabel
+		  kubectl top pods -l name=myLabel`)
+)
+
+var HandledResources []unversioned.GroupKind = []unversioned.GroupKind{
+	api.Kind("Node"),
```

Comment:
> You have flags that only apply to one resource (--containers) and a lot of switch/if conditions because the path is not the same. Also node is more targeted at sysadmins while pod is useful for end users. Having different commands would make it easier if we eventually decide to split this in end-user/admin commands. @deads2k thoughts?

Yes, I think that's probably for the best.  There will probably be a set of common flags and then a series of specific ones.  What we've found on other commands is that flag understability and content assist are improved by subcommands.